File: Backtracking/216. Combination Sum III/solution.cpp
class Solution {
public:
    void dfs(int d, int k, int n, vector<int> curr, vector<vector<int>>& res) {
        if (k==0 && n== 0) {
            res.push_back(curr);
            return;
        }
        if (k <= 0 || n <= 0) {
            return;
        }
        
        for (int i=d+1; i<10; i++) {
            curr.push_back(i);
            dfs(i, k-1, n-i, curr, res);
            curr.pop_back();
        }
        return;
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        dfs(0, k, n, {}, res);
        return res;
    }
};

File: Backtracking/216. Combination Sum III/solution.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def dfs(d, k, n):
            if k == 0 and n == 0:
                return [[]]
            if k <= 0 or n <= 0:
                return None
            
            res = []
            for i in range(d+1, 10):
                arrs = dfs(i, k-1, n-i)
                if arrs is None: continue
                for arr in arrs:
                    arr.append(i)
                res += arrs
            return res
        return dfs(0, k, n)